Sleep-Reduced Thread-Modular Proofs and related CHC refactoring

maul-esel opened the pull request:
This PR contains the integration of sleep set reduction for thread-modular proofs of concurrent programs, as described in our POPL'24 paper. It has been refactored to use the same symbolic independence relations introduced by PR #696.

The PR also contains the refactoring of program-to-CHC translation that was needed to implement the above techique without too much code duplication; and some fixes to our interfaces to CHC solvers.
:arrow_right: @schuessf as you have worked on thread-modular proofs via CHC before, could you / do you want to take a look?

maul-esel committed:
IcfgToChcConcurrent: some documentation and minor refactoring

maul-esel committed:
begin (very hacky) support for sleep sets in Horn clauses

maul-esel committed:
add common interface for variables specific to a thread instance

maul-esel committed:
more simplification with ThreadInstance

maul-esel committed:
new implementation of Horn clause generation for thread-modular proofs

This new implementation is meant to be easier to understand, extensible
and modular. As an example, see the included implementation of sleep set
reduction in the Horn clauses.

The implementation is not yet completely finished, a few TODOs remain.

maul-esel committed:
fix Horn clauses with non-false head

maul-esel committed:
add support for variables not tied to a specific predicate and index

maul-esel committed:
fix location constraint

maul-esel committed:
refactor ChcProviderConcurrentWithLbe: create new ICFG, use existing ChcProvider

maul-esel committed:
add mode and level parameters to ChcProviderConcurrent

maul-esel committed:
add BidirectionalMap::merge (allows to be used in stream collectors)

maul-esel committed:
HcSymbolTable fix copy-paste error

maul-esel committed:
continue refactoring

- ThreadModularHornClauseProvider: give params to individual methods, not constructor
- ThreadModularHornClauseProvider: fix several bugs with generated clauses, e.g. transition constraints
- HornClauseBuilder: default to not modifying variable unless explicitly stated
  (rather than the reverse)
- HornClauseBuilder: support comments on clauses
- ChcProviderConcurrent: use new Horn clause generation

maul-esel committed:
make sleep set vars boolean

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

maul-esel committed:
IcfgToChc: add preferences

maul-esel committed:
extract ConcurrencyMode functionality

maul-esel committed:
refactor Horn clause generation

Definition of the different kinds of clauses, and iteration over the
Icfg are tightly coupled. Hence they are now managed by the same class.

maul-esel committed:
remove obsolete code

maul-esel committed:
add missing license header

maul-esel committed:
implement support for pre-/postconditions

For parametric programs, we can now generate CHC systems that encode
correctness wrt a pre- / postcondition pair. The precondition should
be annotated as a "free requires" on the thread template, Similarly,
the postcondition should be annotated as a "free ensures".

maul-esel committed:
separate ability to specify preconditions from spec mode

maul-esel committed:
IHcReplacementVar: consistently give Script not Sort

maul-esel committed:
add interface for finite-valued variables, and implement for HcSleepVar

maul-esel committed:
PredicateInfo: drop bidirectional map, simplify code

maul-esel committed:
add small optimization: drop Horn clauses with constraint false

maul-esel committed:
add predicate families, to prepare support for different encodings

maul-esel committed:
add settings for future encoding variants

maul-esel committed:
add setting for different preference order encoding

maul-esel committed:
IHcThreadSpecificVar: add forInstance method

maul-esel committed:
fix IcfgToChc preferences, output more detailed error in the future

maul-esel committed:
another fix for settings code

maul-esel committed:
HcSleepVar: properly implement forInstance

maul-esel committed:
SleepSetThreadModularProofs: properly implement symmetry breaking

maul-esel committed:
SleepSetThreadModular w/o symmetry-breaking: simplify code slightly

maul-esel committed:
fix transition constraints for local variables

In non-interference clauses, the interfering thread's local
variables may be updated, but do not appear in the head.
Hence, use body variables instead of head variables for them.

maul-esel committed:
add alternative assert encoding

The assertion condition need not be handled as an assume statement.
Instead, we can include it as constraint in the error clause. This
simplifies the CHC system slightly.

The largest benefit is likely for sleep sets, as we do not need to
consider commutativity wrt the assume statement.

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

# Conflicts:
#	trunk/source/Library-UltimateUtil/src/de/uni_freiburg/informatik/ultimate/util/datastructures/BidirectionalMap.java

schuessf committed:
Fix NPE

schuessf committed:
Fix NPE properly

maul-esel committed:
add some comments

maul-esel committed:
create new package for POR-related classes

maul-esel committed:
add auxiliary method for skippable asserts

maul-esel committed:
fix handling of local variables

Previously, local variables could be read from the wrong thread.
This lead to an unsound encoding. We now only allow reading from
and writing to local variables of the current thread.

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

maul-esel committed:
readable toString for dummy edges (to help debugging)

maul-esel committed:
SemanticIndependenceConditionGenerator: also output raw terms

maul-esel committed:
add SemanticIndependenceConditionGenerator::isSymmetric

maul-esel committed:
HcSymbolTable: add missing mapping of head variables

maul-esel committed:
add support for conditional independence (with precomputed commutativity conditions)

maul-esel committed:
minor preference cleanup: put sleep set settings in container

maul-esel committed:
move CHC categorization to its own class

maul-esel committed:
simpler postcondition encoding: only one thread counter

maul-esel committed:
implement support for different preference orders

For now, sequential composition as well as an approximation of
lockstep are supported.

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

maul-esel committed:
ChcSmtPrinter: use configured file name as prefix for automatic naming

maul-esel committed:
IndependenceChecker: properly substitute global variables

maul-esel committed:
better checking of CHCs

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

maul-esel committed:
fix copy-paste error

maul-esel committed:
add setting to allow semi-commutativity (default: true)

maul-esel committed:
prepare evaluation

maul-esel committed:
fix usage of benchexec variables

maul-esel committed:
add golem during build

maul-esel committed:
TreeAutomizer: must not use Z3 (no longer supports interpolation)

maul-esel committed:
temporarily disable check that crashes UniHorn

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

maul-esel committed:
improve Z3 CHC integration

- ChcSolver: create a fresh Z3 instance
- SmtChcScript: support clauses built with a different script than
    the backend solver, by transferring them if necessary.
    Backtranslation of models is still missing.
- ChcSolver: allow specifying a timeout

maul-esel committed:
SmtChcScript: avoid push/pop

push/pop leads to more UNKNOWN results for Z3. Instead, we use reset.

maul-esel committed:
fix typos in method names

maul-esel committed:
bugfix: record HcBodyVar for each term variable

maul-esel committed:
construct HcHeadVars that reference the proper predicate and index

Our own solvers (UniHorn and TreeAutomizer) heavily rely on the
predicate and index stored in HcPredVars. Thus they fail on the
clauses generated so far. This commit corrects the creation of
HcHeadVars.

maul-esel committed:
eldarica: properly handle scala objects

maul-esel committed:
fix SMT solver setting: TreeAutomizer ignores its own settings

maul-esel committed:
add an IChcScript to invoke eldarica via command line

The current EldaricaChcScript does not work well, we are using the
wrong entry point to eldarica. Until we have figured out the right
one, invoking eldarica via command line will seemingly give better
results.

This is a temporary solution. We should ideally find the right entry
point, or at least some other way that is not this platform-dependent.

maul-esel committed:
log model if found

maul-esel committed:
ChcTransferrer: fix resolution of head variables

maul-esel committed:
ChcTransferrer: implement transferBack(Model)

maul-esel committed:
prepare more extensive evaluation

maul-esel committed:
ChcTransferrer: fix model backtranslation

- only transfer Horn predicate symbols
  (Z3 also gives models for constants corresponding to HcVars)
- fix transferring term variables
- properly transfer function symbol

maul-esel committed:
ApproximateLockstepPreferenceOrder: do not confuse locations with their depth

maul-esel committed:
fix implementation of thread-modular preference orders

This reflects a change in the previously erroneous
(i.e. way too restrictive) definition.

maul-esel committed:
add setting for SKIP_ASSERTION_EDGES

maul-esel committed:
ensure models are in Ultimate normal form

maul-esel committed:
temporarily (for evaluation) avoid let terms in models

maul-esel committed:
SmtChcScript: transfer models to the proper script

maul-esel committed:
throw error for joins even without -ea

maul-esel committed:
parametric programs: support multiple templates and single-instance threads

maul-esel committed:
fix pre/post detection for multiple templates

maul-esel committed:
skip error locations if "skip assert edges" enabled

maul-esel committed:
properly filter error locs from CHC system

maul-esel committed:
fix order for semicommutativity

maul-esel committed:
more work on pre/post for parametric programs

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

maul-esel committed:
temporarily hardcode some conditions for eval purposes

maul-esel committed:
remove unused variable

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

maul-esel committed:
mark unimplemented settings

maul-esel committed:
small settings cleanup

maul-esel committed:
add setting for symmetry clauses

maul-esel committed:
prepare evaluation with symmetry clauses

maul-esel committed:
fix NullPointerException for clause without constraint

maul-esel committed:
HornClause: remove unused ManagedScript parameter

maul-esel committed:
prepare evaluation

maul-esel committed:
prepare evaluation of yet-unsolved tasks

maul-esel committed:
ChcSolver: support passing predicate hints to eldarica

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

# Conflicts:
#	releaseScripts/default/makeZip.sh
#	trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/ChcProviderConcurrentWithLbe.java

maul-esel committed:
add test suite for thread-modular proofs with sleep set reduction

maul-esel committed:
switch to new ISymbolicIndependenceRelation

maul-esel committed:
IndependenceChecker: decouple from HornClauseBuilder

maul-esel committed:
fix assert failure for clauses without head (safety clauses)

maul-esel committed:
fix bug with unknown variables in predicate factory

Move the creation of the symbolic independence relations to
IndependenceChecker, and create a suitable symbol table for
the relations as they work on instantiated conditions.

maul-esel committed:
remove hardcoded commutativity conditions (reverts 5ab1b01)

maul-esel committed:
SleepThreadModularChcTestSuite: specify solver in task definition

maul-esel committed:
remove some redundant tests, fix expected timeout

maul-esel committed:
add setting for nondeterministic sleep updates

maul-esel committed:
prepare evaluation

maul-esel committed:
fix path

maul-esel committed:
fix benchmark file: command line parameters

maul-esel committed:
undo temporary debugging changes

schuessf commented on trunk/source/ChcSolver/src/de/uni_freiburg/informatik/ultimate/plugins/chcsolver/ChcSolverObserver.java:
```diff
@@ -88,11 +101,10 @@ public boolean process(final IElement root) throws Throwable {
 		switch (mPrefs.getBackend()) {
 		case ELDARICA:
-			return new EldaricaChcScript(mServices, annotation.getScript().getScript());
+			// return new EldaricaChcScript(mServices, annotation.getScript().getScript());
```
Can `EldaricaChcScript` be removed now (including the Maven dependencies), and the outcommented code here?

maul-esel replied:
At least currently, it is not being used. If we want to keep using eldarica as a solver long-term (and perhaps on a more fine-granular level than here), we should probably investigate again if we can access it programmatically, such that we can access derivation trees and other, more detailed output.

I'll remove the code for now. We can always go back to it via git later. For the dependencies, see the next comment below.

schuessf commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/EldaricaCliChcScript.java:
```diff
@@ -0,0 +1,207 @@
+	}
+
+	private String getCommand(final File file) {
+		var command = "eld";
```
This only works, if `eld` is in the `PATH` of the system, should we include an Eldarica build in `releaseScripts`?

maul-esel replied:
We need to make this a bit more portable, I agree.

- `eld` is a bash script, so it won't work on Windows. It sets some parameters and then runs a JAR -- we should implement that directly.
- Do we want to check in the eldarica JAR like we do for the SMT solver binaries, or wouldn't it be better to still have it as a Maven dependency?

I just thought about whether we could actually just call eldarica's `main` method from Java, but that might cause all kinds of unexpected behaviour (e.g. memory remaining allocated, re-invocations behaving incorrectly due to old cached stuff, coupling of required java versions), so... maybe not.

schuessf replied:
> Do we want to check in the eldarica JAR like we do for the SMT solver binaries, or wouldn't it be better to still have it as a Maven dependency?

Yes, the Maven dependency is much nicer of course. But I am not sure how you could actually call the binary in that case.
If it is not straightforward to include Eldarica in `releaseScripts`, we might really just assume that `eld` is in the PATH of the system for now (maybe you can throw an useful exception in that case?).
The best solution would be of course to get rid of `EldaricaCliChcScript` again and call Eldarica programmatically. If we cannot get it working, we could include the JAR along with our custom launchers (incl. for Windows) in `releaseScripts`. But this is more of a long term question 😉

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/HcLocalVar.java:
```diff
@@ -39,23 +41,27 @@ public Sort getSort() {
 	public int hashCode() {
-		return Objects.hash(mIndex, mVariable);
+		final int prime = 97;
+		return prime * Objects.hash(mInstance, mVariable);
```
Is there any reason why you multiply the hash code from `Objects::hash` with 97?

maul-esel replied:
Instances of different classes implementing `IHcReplacementVar` are quite likely to end up in the same hash set / map. Even if their internal data is the same, they are not equal and their hashcodes should not collide. Thus, every implementation multiplies its hash code with a different prime number.

I'll add comments explaining this.

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/HcLocationVar.java:
```diff
@@ -36,23 +41,27 @@ public Sort getSort() {
 	public int hashCode() {
-		return Objects.hash(mIndex, mProcedure);
+		final int prime = 79;
+		return prime * Objects.hash(mInstance);
```
Why do you multiply the has code with 79 here?

maul-esel replied:
(see above)

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/IcfgLiptonReducer.java:
```diff
@@ -0,0 +1,185 @@
+ * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
+ *
+ */
+public class IcfgLiptonReducer {
```
Good idea to extract this to an own class 👍
As this this is independent of CHCs, can you move this to some other plugin (maybe Library-IcfgTransformer)?

maul-esel replied:
:+1: I'll have a look

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/ThreadModularHornClauseProvider.java:
```diff
@@ -0,0 +1,843 @@
+
+	// add actual constraints for spec edges, do nothing if not a spec edge
+	protected void transformSpecEdgeClause(final IcfgEdge edge, final HornClauseBuilder clause) {
+		// TODO Fix this for
```
Do you want to address this TODO in this PR?

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/ThreadModularHornClauseProvider.java:
```diff
@@ -0,0 +1,843 @@
+		return mIcfg.getProcedureErrorNodes().entrySet().stream()
+				.flatMap(e -> e.getValue().stream().map(l -> new Pair<>(e.getKey(), l)))
+				.flatMap(e -> mInstances.stream().filter(i -> i.getTemplateName().equals(e.getKey()))
+						.<Pair<ThreadInstance, IcfgLocation>> map(i -> new Pair<>(i, e.getValue())))
```
Is this type parameter necessary here?

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/ThreadModularHornClauseProvider.java:
```diff
@@ -0,0 +1,843 @@
+				.flatMap(e -> e.getValue().getIncomingEdges().stream()
+						.map(t -> new Triple<>(e.getKey(), t.getSource(), t.getTransformula())))
+				.flatMap(e -> mInstances.stream().filter(i -> i.getTemplateName().equals(e.getFirst()))
+						.<Triple<ThreadInstance, IcfgLocation, UnmodifiableTransFormula>> map(
```
Is this type parameter necessary here?

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/ThreadModularHornClauseProvider.java:
```diff
@@ -0,0 +1,843 @@
+		final var clause = createBuilder(mInvariantPredicate, "non-interference clause for transition "
+				+ transition.hashCode() + " with transformula " + transition.getTransformula());
+
+		// TODO support transitions with multiple predecessors (joins)
```
Does this TODO mean that we just ignore joins and thus overapproximate the behavior of the program here?

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/preferences/IcfgToChcPreferenceInitializer.java:
```diff
@@ -0,0 +1,204 @@
+	public static final String LABEL_SYMMETRY_CLAUSES = "Use symmetry clauses";
+	private static final boolean DEF_SYMMETRY_CLAUSES = false;
+
+	// TODO Currently unused
```
Are there any plans to use the settings right now?

schuessf commented on trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/preferences/IcfgToChcPreferenceInitializer.java:
```diff
@@ -0,0 +1,204 @@
+			It is particularly recommended in connection with necessary-and-complete conditional independence.""";
+	private static final boolean DEF_NONDET_SLEEP_UPDATES = false;
+
+	// TODO Currently unused
```
Are there any plans to use the settings right now?

schuessf commented on trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/ChcSolutionNormalizer.java:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	public Derivation normalize(final Derivation derivation) {
+		// TODO
```
Is the normalization relevant for now?

schuessf reviewed (commented):
Thanks for your work! Your changes drastically improve the code (that was rather ad-hoc by me 😄). I have just a few comments left.

maul-esel committed:
remove EldaricaChcScript and related classes (for now)

Currently, the implementation does not work well: CHC systems that are
easily solved by eldarica via the command line cannot be solved by this
implementation. We are probably using the eldarica classes incorrectly.

It's also not clear that eldarica provides a stable API, we are just
accessing some implementation details here that might change at any time.

maul-esel committed:
improvements to IHcReplacementVar implementations

- document usage of different prime numbers for distinct hash codes
- document purpose of classes
- make implementations final
- additional integrity checks (null checks, refined types of program variables)
- updated & simplified equals()

maul-esel committed:
add missing benchmark rundefinitions

maul-esel committed:
Merge branch 'dev' into wip/dk/sleep-threadmodular

# Conflicts:
#	.gitignore
#	trunk/source/ChcSolver/src/de/uni_freiburg/informatik/ultimate/plugins/chcsolver/ChcSolverObserver.java
#	trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/HcGlobalVar.java
#	trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/HcLocalVar.java
#	trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/HcLocationVar.java
#	trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/IcfgToChcConcurrent.java
#	trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/preferences/IcfgToChcPreferences.java
#	trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/HcBodyVar.java
#	trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/HcPredVar.java
#	trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/HornUtilConstants.java
#	trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/eldarica/Backtranslator.java

maul-esel committed:
CC fixes

maul-esel committed:
update to golem 0.9.0

maul-esel committed:
new file structure for SleepThreadModularChcTestSuite and benchexec rundefs

maul-esel committed:
IcfgToChcPreferenceInitializer: make sleep set reduction opt-in

maul-esel committed:
parameterized program task definitions: fix relative paths

maul-esel committed:
ThreadModularHornClauseProvider: fix symmetry-clause implementation

The old implementation made the classic mistake of implementing the
substitution of multiple variables by repeatedly performing single-
variable substitutions of each variable individually.
As the substitutions here are cyclic (being permutations, in fact),
this caused later substitutions to overwrite earlier substitutions.
Depending on nondeterministic iteration order, this produced either
trivially satisfiable or trivially unsatisfiable CHC systems.

maul-esel committed:
fix rundefinition: inconsistent parameter sets for some backends

maul-esel committed:
correct wording in settings description

maul-esel committed:
add tests for nondet-sleep

maul-esel committed:
ChcSolver: add option to dump solution as eldarica hints
